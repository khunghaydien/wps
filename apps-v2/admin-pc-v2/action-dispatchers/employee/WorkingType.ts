import { bindActionCreators } from 'redux';

import {
  catchApiError,
  loadingEnd,
  loadingStart,
} from '@apps/commons/actions/app';
import msg from '@apps/commons/languages';
import TextUtil from '@apps/commons/utils/TextUtil';

import WorkingTypeRepository, {
  ORDER_TYPE,
  SortOrder,
} from '@apps/repositories/organization/workingType/WorkingTypeRepository';

import {
  actions as listActions,
  initialState,
} from '@admin-pc/modules/workingType/ui/list';
import { actions as PagingActions } from '@admin-pc/modules/workingType/ui/paging';
import {
  actions as searchActions,
  Condition,
  initCondition as $initCondition,
  WORK_SYSTEM_TYPE,
} from '@admin-pc/modules/workingType/ui/searchCondition';
import {
  actions as workingTypeDialogActions,
  WorkingType,
} from '@admin-pc-v2/modules/employee/ui/workingTypeDialog';

import { AppDispatch } from '@admin-pc/action-dispatchers/AppThunk';
import {
  SEARCH_WORKING_TYPE,
  searchWorkingType,
} from '@admin-pc/actions/workingType';

type SearchCondition = Condition & { companyId: string };

const convertCondition = (searchCondition: SearchCondition) => {
  const { code, ...condition } = searchCondition;
  return {
    ...condition,
    id: null,
    codes: code ? [code] : [],
    isSearchCodeForPartialMatch: true,
  };
};

export const convertWorkSystem = (workSystem) => {
  if (workSystem === WORK_SYSTEM_TYPE.JP_Flex_Without_Core) {
    return WORK_SYSTEM_TYPE.JP_Flex;
  }
  return workSystem;
};

interface PatternListService {
  init: (companyId: string, targetDate: string, chunkSize: number) => void;
  search: (
    searchCondition: SearchCondition,
    sortOrder: SortOrder,
    chunkSize: number,
    changeSort: boolean
  ) => void;
  turnPage: (
    searchCondition: SearchCondition,
    sortOrder: SortOrder,
    chunkSize: number,
    offsetCode: string,
    currentPage: number
  ) => void;
  setSearchCondition: (key: string, value: string) => void;
  saveCondition: (condition: SearchCondition) => void;
  onClickRow: (workingType: WorkingType) => void;
  openDialog: (targetDate: string, companyId: string) => void;
  closeDiaglog: () => void;
  resetWorkingType: () => void;
}

export default (dispatch: AppDispatch): PatternListService => {
  const searchCondition = bindActionCreators(searchActions, dispatch);
  const list = bindActionCreators(listActions, dispatch);
  const paging = bindActionCreators(PagingActions, dispatch);
  const workingTypeDialog = bindActionCreators(
    workingTypeDialogActions,
    dispatch
  );

  const conditionIsNull = (value: string, name: string): boolean => {
    if (!value || value === '') {
      dispatch(
        catchApiError(
          {
            name: 'conditionIsNull',
            message: TextUtil.template(msg().Com_Err_Specify, name),
          },
          { isContinuable: true }
        )
      );
      return true;
    }
    return false;
  };

  return {
    init: (companyId: string, targetDate: string, chunkSize: number) => {
      const initCondition = {
        ...$initCondition,
        companyId,
        targetDate,
      };

      searchCondition.initialize({
        origin: initCondition,
        new: initCondition,
      });
      list.initialize();

      dispatch(loadingStart());
      const condition = convertCondition(initCondition);
      WorkingTypeRepository.searchOffsetCodes({
        condition,
        sortOrder: initialState.sortOrder,
        chunkSize,
      })
        .then(({ total, offsetCodes, hasMoreRecords }) => {
          dispatch(loadingEnd());
          list.setOffsetCodes(total, offsetCodes, hasMoreRecords);
          if (total > 0) {
            paging.setCurrent(1);
          }
        })
        .catch((err) => {
          dispatch(loadingEnd());
          dispatch(catchApiError(err, { isContinuable: true }));
        });

      dispatch(
        searchWorkingType({
          condition,
          sortOrder: initialState.sortOrder,
          chunkSize,
          offsetCode: null,
        })
      );
    },
    search: async (
      searchCondition: SearchCondition,
      sortOrder: SortOrder,
      chunkSize: number,
      changeSort: boolean
    ) => {
      let $sortOrder = sortOrder;

      if (
        conditionIsNull(searchCondition.targetDate, msg().Admin_Lbl_TargetDate)
      ) {
        return;
      }

      if (changeSort) {
        $sortOrder = {
          field: $sortOrder.field,
          order:
            $sortOrder.order === ORDER_TYPE.ASC
              ? ORDER_TYPE.DESC
              : ORDER_TYPE.ASC,
        };
        list.setSortOrder($sortOrder);
      } else {
        $sortOrder = {
          field: $sortOrder.field,
          order: ORDER_TYPE.ASC,
        };
        list.resetSortOrder();
      }

      dispatch(loadingStart());
      const condition = {
        ...convertCondition(searchCondition),
        workSystem: convertWorkSystem(searchCondition.workSystem),
      };
      await WorkingTypeRepository.searchOffsetCodes({
        condition,
        sortOrder: $sortOrder,
        chunkSize,
      })
        .then(({ total, offsetCodes, hasMoreRecords }) => {
          dispatch(loadingEnd());
          list.setOffsetCodes(total, offsetCodes, hasMoreRecords);
          if (total > 0) {
            paging.setCurrent(1);
          } else {
            paging.setCurrent(0);
          }
        })
        .catch((err) => {
          dispatch(loadingEnd());
          dispatch(catchApiError(err, { isContinuable: true }));
        });
      dispatch(
        searchWorkingType({
          condition,
          sortOrder: $sortOrder,
          chunkSize,
          offsetCode: null,
        })
      );
    },
    turnPage: (
      searchCondition: SearchCondition,
      sortOrder: SortOrder,
      chunkSize: number,
      offsetCode: string,
      currentPage: number
    ) => {
      const condition = {
        ...convertCondition(searchCondition),
        workSystem: convertWorkSystem(searchCondition.workSystem),
      };
      dispatch(
        searchWorkingType({
          condition,
          sortOrder,
          chunkSize,
          offsetCode,
        })
      );
      dispatch(paging.setCurrent(currentPage));
    },
    setSearchCondition: searchCondition.setNew,
    saveCondition: searchCondition.setOrigin,
    onClickRow: workingTypeDialog.setSelectedWorkingType,
    openDialog: (targetDate: string, companyId: string) => {
      if (
        conditionIsNull(targetDate, msg().Admin_Lbl_ValidDateFrom) ||
        conditionIsNull(companyId, msg().Admin_Lbl_Company)
      ) {
        return;
      }

      workingTypeDialog.show();
    },
    closeDiaglog: () => {
      workingTypeDialog.hide();
      searchCondition.initialize({
        origin: $initCondition,
        new: $initCondition,
      });
      paging.reset();
      list.initialize();
      dispatch({ type: SEARCH_WORKING_TYPE, payload: [] });
      workingTypeDialog.resetSelectedWorkingType();
    },
    resetWorkingType: () => {
      dispatch({ type: SEARCH_WORKING_TYPE, payload: [] });
    },
  };
};
