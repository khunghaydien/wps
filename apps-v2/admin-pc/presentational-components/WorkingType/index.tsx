import React from 'react';

import cloneDeep from 'lodash/cloneDeep';

import configList from '../../constants/configList/workingType';

import { Action, Record } from '../../utils/RecordUtil';

import MainContents from './MainContents';

export type Props = {
  actions: Action;
  companyId: string;
  searchWorkingType: Array<Record>;
  tmpEditRecordHistory: Record;
};

export default class WorkingType extends React.Component<Props> {
  constructor(props) {
    super(props);

    const param = { companyId: props.companyId };
    this.props.actions.searchLeave(param);
    this.props.actions.searchObjectivelyEventLogSetting(param);
    this.props.actions.searchRestReason(param);
    this.props.actions.searchAllowance(param);
    this.props.actions.searchLateArrivalEarlyLeaveReason(param);
    this.props.actions.getConstantsWorkingType();
    this.props.actions.searchRecordExtendedItemSet(param);
    this.props.actions.searchRecordDisplayFieldLayout(param);
    this.props.actions.searchAttExtendedItem(param);
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    if (this.props.companyId !== nextProps.companyId) {
      const param = { companyId: nextProps.companyId };
      this.props.actions.searchLeave(param);
      this.props.actions.searchObjectivelyEventLogSetting(param);
      this.props.actions.searchRestReason(param);
      this.props.actions.searchAllowance(param);
      this.props.actions.searchLateArrivalEarlyLeaveReason(param);
      this.props.actions.searchRecordExtendedItemSet(param);
      this.props.actions.searchRecordDisplayFieldLayout(param);
      this.props.actions.searchAttExtendedItem(param);
    }
  }

  render() {
    const configListWorkingType = cloneDeep(configList);
    configListWorkingType.base.forEach((config) => {
      if (config.key === 'companyId') {
        config.defaultValue = this.props.companyId;
      }
    });
    return (
      <MainContents
        // @ts-ignore
        componentKey="workingType"
        configList={configListWorkingType}
        itemList={this.props.searchWorkingType}
        {...this.props}
      />
    );
  }
}
