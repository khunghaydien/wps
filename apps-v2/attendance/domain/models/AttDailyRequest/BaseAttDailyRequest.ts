import { compose } from '../../../../commons/utils/FnUtil';

// FIXME: 本来は、 model で repository の実態を呼び出すべきではないが、 createFromRemote が repository に移行するまでは許可することにする
import { DailyRest } from '@attendance/repositories/models/DailyRecord';
// FIXME: 本来は、 model で repository の実態を呼び出すべきではないが、 createFromRemote が repository に移行するまでは許可することにする
import { Timesheet as TimesheetFromRemote } from '@attendance/repositories/models/Timesheet';

import $STATUS from '@apps/domain/models/approval/request/Status';
import { RestTimeReason as DomainRestTimeReason } from '@attendance/domain/models/RestTimeReason';

import {
  CODE as REQUEST_TYPE_CODE,
  Code as AttDailyRequestTypeCode,
  DailyRequestNameMap,
} from '../AttDailyRequestType';
import { LeaveRange } from '../LeaveRange';
import { LeaveType } from '../LeaveType';
import * as RestTime from '../RestTime';
import { SubstituteLeaveType } from '../SubstituteLeaveType';
// FIXME: 本来は、model で service は呼び出すべきではないが、createDefaultValue が service に移行するまでは許可することにする
import getRequestTypeName from '@apps/attendance/domain/services/DailyRequestService/getRequestTypeName';
// FIXME: 本来は、model で factory は呼び出すべきではないが、createDefaultValue が service に移行するまでは許可することにする
import $createRestTimesFactory from '@attendance/domain/factories/RestTimesFactory';

const createRestTimesFactory = $createRestTimesFactory();

export const STATUS = {
  NOT_REQUESTED: $STATUS.NotRequested,
  RECALLED: $STATUS.Recalled,
  REJECTED: $STATUS.Rejected,
  APPROVAL_IN: $STATUS.ApprovalIn,
  APPROVED: $STATUS.Approved,
  CANCELED: $STATUS.Canceled,
  REAPPLYING: $STATUS.Reapplying,
} as const;

export type Status = Value<typeof STATUS>;

export type AttDailyRequestFromRemote = Readonly<{
  // 申請ID
  id: string;
  // 申請タイプコード
  requestTypeCode: AttDailyRequestTypeCode;
  // 申請ステータス
  status: Status;
  // 開始日(YYYY-MM-DD)
  startDate: string;
  // 終了日(YYYY-MM-DD)
  endDate: string;
  // 開始時刻
  startTime: number | null;
  // 終了時刻
  endTime: number | null;
  // 備考	最大255文字。
  remarks: string | null;
  // 理由 最大255文字。
  reason: string | null;
  // 休暇名
  leaveName: string | null;
  // 休暇コード
  leaveCode: string | null;
  // 休暇タイプ
  leaveType: LeaveType | null;
  // 休暇範囲
  leaveRange: LeaveRange | null;
  // 休暇内訳名
  leaveDetailName: string | null;
  // 休暇内訳コード
  leaveDetailCode: string | null;
  // 休日出勤取得休日タイプ
  substituteLeaveType: SubstituteLeaveType | null;
  // 振替休日取得日
  substituteDate: string | null;
  // 理由を求めるか否かフラグ
  requireReason: boolean;
  // 再申請元申請ID
  originalRequestId: string | null;
  // 承認内容変更申請中フラグ
  isForReapply: boolean;
  // 直行直帰申請休憩リスト
  directApplyRestList: DailyRest[];
  // 勤務パターンコード
  patternCode: string | null;
  // 勤務パターン名
  patternName: string | null;
  // 勤務時間変更申請の休憩開始時間1のデフォルト値
  patternApplyRest1StartTime: number | null;
  // 勤務時間変更申請の休憩終了時間1のデフォルト値
  patternApplyRest1EndTime: number | null;
  // 勤務時間変更申請の休憩開始時間2のデフォルト値
  patternApplyRest2StartTime: number | null;
  // 勤務時間変更申請の休憩終了時間2のデフォルト値
  patternApplyRest2EndTime: number | null;
  // 勤務時間変更申請の休憩開始時間3のデフォルト値
  patternApplyRest3StartTime: number | null;
  // 勤務時間変更申請の休憩終了時間3のデフォルト値
  patternApplyRest3EndTime: number | null;
  // 勤務時間変更申請の休憩開始時間4のデフォルト値
  patternApplyRest4StartTime: number | null;
  // 勤務時間変更申請の休憩終了時間4のデフォルト値
  patternApplyRest4EndTime: number | null;
  // 勤務時間変更申請の休憩開始時間5のデフォルト値
  patternApplyRest5StartTime: number | null;
  // 勤務時間変更申請の休憩終了時間2のデフォルト値
  patternApplyRest5EndTime: number | null;
  // 承認者01の名前
  approver01Name: string;
  // 自己都合
  personalReason: boolean;
  // 自責管理
  useManagePersonalReason: boolean;
  // 労働時間制
  workSystem: string;
  // フレックス開始時刻
  flexStartTime: number | null;
  // フレックス終了時刻
  flexEndTime: number | null;
  // コアタイムなし
  withoutCoreTime: boolean;
  // 勤務変更の日タイプ
  requestDayType: string;
  // 直接入力
  isDirectInputTimeRequest: boolean;
  // 遅刻早退理由D
  lateArrivalEarlyLeaveReasonId: string;
  // 遅刻早退理由名
  lateArrivalEarlyLeaveReasonName: string;
  // 遅刻早退理由コード
  lateArrivalEarlyLeaveReasonCode: string;
}>;

export type BaseAttDailyRequest = {
  // 申請ID
  id: string;
  // 申請タイプコード
  requestTypeCode: AttDailyRequestTypeCode;
  // 申請種別名
  requestTypeName: string;
  // 申請ステータス
  status: Status;
  // 開始日(YYYY-MM-DD)
  startDate: string;
  // 終了日(YYYY-MM-DD)
  endDate: string;
  // 開始時刻
  startTime: number | null;
  // 終了時刻
  endTime: number | null;
  // 備考	最大255文字。
  remarks: string;
  // 理由 最大255文字。
  reason: string;
  // 休暇コード
  leaveCode: string | null;
  // 休暇名
  leaveName: string | null;
  // 休暇タイプ
  leaveType: LeaveType | null;
  // 休暇範囲
  leaveRange: LeaveRange | null;
  // 休暇内訳名
  leaveDetailName: string | null;
  // 休暇内訳コード
  leaveDetailCode: string | null;
  // 休日出勤取得休日タイプ
  substituteLeaveType: SubstituteLeaveType | null;
  // 振替休日取得日
  substituteDate: string | null;
  // 直行直帰申請休憩
  directApplyRestTimes: RestTime.RestTimes;
  // 勤務パターンコード
  patternCode: string | null;
  // 勤務パターン名
  patternName: string | null;
  // 勤務時間変更申請休憩
  patternRestTimes: RestTime.RestTimes;
  // 再申請元申請ID
  originalRequestId: string | null;
  // 承認内容変更申請中フラグ
  isForReapply: boolean;
  // 理由を求めるか否かフラグ
  requireReason: boolean;
  // 承認者01の名前
  approver01Name: string;
  // 自己都合
  personalReason: boolean;
  // 自責管理
  useManagePersonalReason: boolean;
  // 労働時間制
  workSystem: string;
  // フレックス開始時刻
  flexStartTime: number | null;
  // フレックス終了時刻
  flexEndTime: number | null;
  // コアタイムなし
  withoutCoreTime: boolean;
  // 勤務変更の日タイプ
  requestDayType: string | null;
  // 申請可能な勤務変更の日タイプ
  requestableDayType: string | null;
  // 直接入力
  isDirectInputTimeRequest: boolean;
  // 申請可能な直接入力
  canDirectInputTimeRequest: boolean;
  // 遅刻早退理由ID
  reasonId: string;
  // 遅刻早退理由名
  reasonName: string;
  // 遅刻早退理由コード
  reasonCode: string;
  // 遅刻理由を管理するか否か
  useLateArrivalReason: boolean;
  // 早退理由を管理するか否か
  useEarlyLeaveReason: boolean;
};

export type IBaseAttDailyRequestFactory<T extends BaseAttDailyRequest> = {
  create: (...args: unknown[]) => T | Promise<T>;
};

const mapDirectApplyRestTime = (
  fromRemote: AttDailyRequestFromRemote
): {
  startTime: number;
  endTime: number;
  restReason: DomainRestTimeReason | null;
}[] =>
  (fromRemote.directApplyRestList || []).map((rest) => ({
    startTime: rest.restStartTime,
    endTime: rest.restEndTime,
    restReason: rest.restReason,
  }));

const createDirectApplyRestTimesFromRemote = (
  fromRemote: AttDailyRequestFromRemote
): RestTime.RestTimes => {
  const RestTimesFactory = createRestTimesFactory();
  return compose(
    RestTimesFactory.filter,
    RestTime.convertRestTimes,
    mapDirectApplyRestTime
  )(fromRemote);
};

const mapPatternRestTime = (
  fromRemote: AttDailyRequestFromRemote
): {
  startTime: any;
  endTime: any;
}[] => [
  {
    startTime: fromRemote.patternApplyRest1StartTime,
    endTime: fromRemote.patternApplyRest1EndTime,
  },
  {
    startTime: fromRemote.patternApplyRest2StartTime,
    endTime: fromRemote.patternApplyRest2EndTime,
  },
  {
    startTime: fromRemote.patternApplyRest3StartTime,
    endTime: fromRemote.patternApplyRest3EndTime,
  },
  {
    startTime: fromRemote.patternApplyRest4StartTime,
    endTime: fromRemote.patternApplyRest4EndTime,
  },
  {
    startTime: fromRemote.patternApplyRest5StartTime,
    endTime: fromRemote.patternApplyRest5EndTime,
  },
];

const createPatternRestTimesFromRemote = (
  fromRemote: AttDailyRequestFromRemote
): RestTime.RestTimes =>
  compose(RestTime.convertRestTimes, mapPatternRestTime)(fromRemote);

export const createFromRemote = (
  attRequestTypeMap: DailyRequestNameMap,
  paramFromRemote: AttDailyRequestFromRemote,
  timesheet: TimesheetFromRemote
): BaseAttDailyRequest => {
  const dailyRecord = timesheet.records.find(
    ({ recordDate }) => recordDate === paramFromRemote.startDate
  );
  const requestTypeName = getRequestTypeName(paramFromRemote.requestTypeCode, {
    nameMap: attRequestTypeMap,
    dailyRecord,
  });
  return {
    id: paramFromRemote.id,
    requestTypeCode: paramFromRemote.requestTypeCode,
    requestTypeName,
    status: paramFromRemote.status,
    startDate: paramFromRemote.startDate,
    endDate: paramFromRemote.endDate,
    startTime: paramFromRemote.startTime,
    endTime: paramFromRemote.endTime,
    remarks: paramFromRemote.remarks || '',
    reason: paramFromRemote.reason || '',
    leaveName: paramFromRemote.leaveName,
    leaveCode: paramFromRemote.leaveCode,
    leaveType: paramFromRemote.leaveType,
    leaveRange: paramFromRemote.leaveRange,
    leaveDetailCode: paramFromRemote.leaveDetailCode,
    leaveDetailName: paramFromRemote.leaveDetailName,
    substituteLeaveType: paramFromRemote.substituteLeaveType,
    substituteDate: paramFromRemote.substituteDate,
    directApplyRestTimes: createDirectApplyRestTimesFromRemote(paramFromRemote),
    patternCode: paramFromRemote.patternCode,
    patternName: paramFromRemote.patternName,
    patternRestTimes: createPatternRestTimesFromRemote(paramFromRemote),
    requireReason: paramFromRemote.requireReason,
    originalRequestId: paramFromRemote.originalRequestId,
    isForReapply: paramFromRemote.isForReapply,
    approver01Name: paramFromRemote.approver01Name,
    personalReason: paramFromRemote.personalReason,
    useManagePersonalReason: paramFromRemote.useManagePersonalReason,
    workSystem: paramFromRemote.workSystem,
    flexStartTime: paramFromRemote.flexStartTime,
    flexEndTime: paramFromRemote.flexEndTime,
    withoutCoreTime: paramFromRemote.withoutCoreTime,
    requestDayType: paramFromRemote.requestDayType,
    requestableDayType: null,
    isDirectInputTimeRequest: paramFromRemote.isDirectInputTimeRequest,
    canDirectInputTimeRequest: false,
    reasonId: paramFromRemote.lateArrivalEarlyLeaveReasonId,
    reasonName: paramFromRemote.lateArrivalEarlyLeaveReasonName,
    reasonCode: paramFromRemote.lateArrivalEarlyLeaveReasonCode,
    useLateArrivalReason: false,
    useEarlyLeaveReason: false,
  };
};

export const defaultValue: BaseAttDailyRequest = {
  id: '',
  requestTypeCode: REQUEST_TYPE_CODE.None,
  requestTypeName: '',
  status: STATUS.NOT_REQUESTED,
  startDate: '',
  endDate: '',
  startTime: null,
  endTime: null,
  remarks: '',
  reason: '',
  leaveName: '',
  leaveCode: '',
  leaveType: null,
  leaveRange: null,
  leaveDetailCode: null,
  leaveDetailName: null,
  substituteLeaveType: null,
  substituteDate: '',
  originalRequestId: '',
  isForReapply: false,
  directApplyRestTimes: [],
  patternCode: null,
  patternName: null,
  patternRestTimes: [],
  requireReason: false,
  approver01Name: '',
  personalReason: false,
  useManagePersonalReason: false,
  workSystem: '',
  flexStartTime: null,
  flexEndTime: null,
  withoutCoreTime: false,
  requestDayType: null,
  requestableDayType: null,
  isDirectInputTimeRequest: false,
  canDirectInputTimeRequest: false,
  reasonId: null,
  reasonName: null,
  reasonCode: null,
  useLateArrivalReason: false,
  useEarlyLeaveReason: false,
};

export const createFromDefaultValue = (
  attRequestTypeMap: DailyRequestNameMap,
  requestTypeCode: AttDailyRequestTypeCode,
  value: BaseAttDailyRequest = defaultValue
) => ({
  ...value,
  requestTypeCode: attRequestTypeMap[requestTypeCode].code,
  requestTypeName: attRequestTypeMap[requestTypeCode].name,
});
