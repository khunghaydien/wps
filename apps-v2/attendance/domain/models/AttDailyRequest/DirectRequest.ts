import { compose } from '../../../../commons/utils/FnUtil';

import { CODE } from '../AttDailyRequestType';
import * as RestTime from '../RestTime';
import {
  createDirectApplyRestTimes,
  getDirectApplyTimeRange,
  WorkingType,
} from '../WorkingType';
import { BaseAttDailyRequest } from './BaseAttDailyRequest';

export type DirectRequest = BaseAttDailyRequest & {
  type: typeof CODE.Direct;

  /**
   * 開始日
   */
  startDate: string;

  /**
   * 終了日
   */
  endDate: string;

  /**
   * 開始時間
   */
  startTime: number | null;

  /**
   * 終了時間
   */
  endTime: number | null;

  /**
   * 休憩時間
   */
  directApplyRestTimes: RestTime.RestTimes;
};

/**
 * AttDailyRequest から DirectRequest を作成します。
 */
const formatAttDailyRequest = (
  request: BaseAttDailyRequest
): DirectRequest => ({
  ...request,
  type: CODE.Direct,
});

/**
 *  開始日と終了日のデフォルト値の設定します。
 */
const defaultStartDateAndEndDate =
  (targetDate: string | null) => (request: DirectRequest) => {
    if (!targetDate || request.startDate) {
      return request;
    }
    return { ...request, startDate: targetDate, endDate: targetDate };
  };

/**
 * 勤務体系から基本設定を作成します。
 */
const defaultValue =
  (workingType: WorkingType | null = null) =>
  (request: DirectRequest): DirectRequest => {
    if (request.id || !workingType) {
      return request;
    }
    return {
      ...request,
      ...getDirectApplyTimeRange(workingType),
      directApplyRestTimes: createDirectApplyRestTimes(workingType),
    };
  };

export const update = (
  target: DirectRequest,
  key: string,
  value: DirectRequest[keyof DirectRequest]
): DirectRequest => ({
  ...target,
  [key]: value,
});

export const create = (
  request: BaseAttDailyRequest,
  workingType: WorkingType | null = null,
  targetDate: string | null = null
): DirectRequest =>
  compose(
    defaultStartDateAndEndDate(targetDate),
    defaultValue(workingType),
    formatAttDailyRequest
  )(request);
