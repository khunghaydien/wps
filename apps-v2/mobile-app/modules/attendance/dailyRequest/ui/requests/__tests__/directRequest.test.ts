import snapshotDiff from 'snapshot-diff';

import { MAX_STANDARD_REST_TIME_COUNT } from '@apps/attendance/domain/models/RestTime';
import { defaultValue } from '@attendance/domain/models/AttDailyRequest/BaseAttDailyRequest';
import { create } from '@attendance/domain/models/AttDailyRequest/DirectRequest';
import { defaultValue as workingType } from '@attendance/domain/models/WorkingType';

import reducer, {
  actions,
  CLEAR,
  INITIALIZE,
  initialState,
  UPDATE,
} from '../directRequest';

jest.mock('nanoid', () => {
  let i = 1;
  return {
    __esModule: true,
    default: () => `${i++}`,
  };
});

describe('reducer()', () => {
  it('@@init', () => {
    // @ts-ignore
    const next = reducer(undefined, { type: '@@INIT' });
    expect(snapshotDiff({}, next)).toMatchSnapshot();
  });
  describe(`${INITIALIZE}`, () => {
    it('should be executed', () => {
      const request = create(defaultValue, workingType, '2022-02-22');
      const next = reducer(
        initialState,
        actions.initialize(request, MAX_STANDARD_REST_TIME_COUNT)
      );
      expect(snapshotDiff(initialState, next)).toMatchSnapshot();
    });
  });
  describe(`${UPDATE}`, () => {
    it('should not be executed if state is not initializing', () => {
      const next = reducer(initialState, actions.update('startTime', 9 * 60));
      expect(snapshotDiff(initialState, next)).toMatchSnapshot();
    });
    it('should be executed ', () => {
      // @ts-ignore
      const request = create(defaultValue, workingType, '2022-02-22');
      const state = reducer(
        initialState,
        actions.initialize(request, MAX_STANDARD_REST_TIME_COUNT)
      );
      const next = reducer(state, actions.update('startTime', 9 * 60));
      expect(snapshotDiff(state, next)).toMatchSnapshot();
    });
  });
  describe(`${CLEAR}`, () => {
    it('should be executed', () => {
      // @ts-ignore
      const request = create(defaultValue, workingType, '2022-02-22');
      const state = reducer(
        initialState,
        actions.initialize(request, MAX_STANDARD_REST_TIME_COUNT)
      );
      const next = reducer(state, actions.clear());
      expect(snapshotDiff(state, next)).toMatchSnapshot();
    });
  });
});
