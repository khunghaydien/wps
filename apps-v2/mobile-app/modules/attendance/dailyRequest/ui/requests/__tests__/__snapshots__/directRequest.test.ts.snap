// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer() @@init 1`] = `
"Snapshot Diff:
- First value
+ Second value

- Object {}
+ Object {
+   \\"request\\": null,
+ }"
`;

exports[`reducer() MOBILE_APP/MODULES/ATTENDANCE/DAILY_REQUEST/UI/REQUESTS/DIRECT_REQUEST/CLEAR should be executed 1`] = `
"Snapshot Diff:
- First value
+ Second value

  Object {
-   \\"request\\": Object {
-     \\"approver01Name\\": \\"\\",
-     \\"canDirectInputTimeRequest\\": false,
-     \\"directApplyRestTimes\\": Array [
-       Object {
-         \\"endTime\\": null,
-         \\"id\\": \\"3\\",
-         \\"restReason\\": null,
-         \\"startTime\\": null,
-       },
-     ],
-     \\"endDate\\": \\"2022-02-22\\",
-     \\"endTime\\": null,
-     \\"flexEndTime\\": null,
-     \\"flexStartTime\\": null,
-     \\"id\\": \\"\\",
-     \\"isDirectInputTimeRequest\\": false,
-     \\"isForReapply\\": false,
-     \\"leaveCode\\": \\"\\",
-     \\"leaveDetailCode\\": null,
-     \\"leaveDetailName\\": null,
-     \\"leaveName\\": \\"\\",
-     \\"leaveRange\\": null,
-     \\"leaveType\\": null,
-     \\"originalRequestId\\": \\"\\",
-     \\"patternCode\\": null,
-     \\"patternName\\": null,
-     \\"patternRestTimes\\": Array [],
-     \\"personalReason\\": false,
-     \\"reason\\": \\"\\",
-     \\"reasonCode\\": null,
-     \\"reasonId\\": null,
-     \\"reasonName\\": null,
-     \\"remarks\\": \\"\\",
-     \\"requestDayType\\": null,
-     \\"requestTypeCode\\": \\"None\\",
-     \\"requestTypeName\\": \\"\\",
-     \\"requestableDayType\\": null,
-     \\"requireReason\\": false,
-     \\"startDate\\": \\"2022-02-22\\",
-     \\"startTime\\": null,
-     \\"status\\": \\"NotRequested\\",
-     \\"substituteDate\\": \\"\\",
-     \\"substituteLeaveType\\": null,
-     \\"type\\": \\"Direct\\",
-     \\"useEarlyLeaveReason\\": false,
-     \\"useLateArrivalReason\\": false,
-     \\"useManagePersonalReason\\": false,
-     \\"withoutCoreTime\\": false,
-     \\"workSystem\\": \\"\\",
-   },
+   \\"request\\": null,
  }"
`;

exports[`reducer() MOBILE_APP/MODULES/ATTENDANCE/DAILY_REQUEST/UI/REQUESTS/DIRECT_REQUEST/INITIALIZE should be executed 1`] = `
"Snapshot Diff:
- First value
+ Second value

  Object {
-   \\"request\\": null,
+   \\"request\\": Object {
+     \\"approver01Name\\": \\"\\",
+     \\"canDirectInputTimeRequest\\": false,
+     \\"directApplyRestTimes\\": Array [
+       Object {
+         \\"endTime\\": null,
+         \\"id\\": \\"1\\",
+         \\"restReason\\": null,
+         \\"startTime\\": null,
+       },
+     ],
+     \\"endDate\\": \\"2022-02-22\\",
+     \\"endTime\\": null,
+     \\"flexEndTime\\": null,
+     \\"flexStartTime\\": null,
+     \\"id\\": \\"\\",
+     \\"isDirectInputTimeRequest\\": false,
+     \\"isForReapply\\": false,
+     \\"leaveCode\\": \\"\\",
+     \\"leaveDetailCode\\": null,
+     \\"leaveDetailName\\": null,
+     \\"leaveName\\": \\"\\",
+     \\"leaveRange\\": null,
+     \\"leaveType\\": null,
+     \\"originalRequestId\\": \\"\\",
+     \\"patternCode\\": null,
+     \\"patternName\\": null,
+     \\"patternRestTimes\\": Array [],
+     \\"personalReason\\": false,
+     \\"reason\\": \\"\\",
+     \\"reasonCode\\": null,
+     \\"reasonId\\": null,
+     \\"reasonName\\": null,
+     \\"remarks\\": \\"\\",
+     \\"requestDayType\\": null,
+     \\"requestTypeCode\\": \\"None\\",
+     \\"requestTypeName\\": \\"\\",
+     \\"requestableDayType\\": null,
+     \\"requireReason\\": false,
+     \\"startDate\\": \\"2022-02-22\\",
+     \\"startTime\\": null,
+     \\"status\\": \\"NotRequested\\",
+     \\"substituteDate\\": \\"\\",
+     \\"substituteLeaveType\\": null,
+     \\"type\\": \\"Direct\\",
+     \\"useEarlyLeaveReason\\": false,
+     \\"useLateArrivalReason\\": false,
+     \\"useManagePersonalReason\\": false,
+     \\"withoutCoreTime\\": false,
+     \\"workSystem\\": \\"\\",
+   },
  }"
`;

exports[`reducer() MOBILE_APP/MODULES/ATTENDANCE/DAILY_REQUEST/UI/REQUESTS/DIRECT_REQUEST/UPDATE should be executed  1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -37,11 +37,11 @@
      \\"requestTypeCode\\": \\"None\\",
      \\"requestTypeName\\": \\"\\",
      \\"requestableDayType\\": null,
      \\"requireReason\\": false,
      \\"startDate\\": \\"2022-02-22\\",
-     \\"startTime\\": null,
+     \\"startTime\\": 540,
      \\"status\\": \\"NotRequested\\",
      \\"substituteDate\\": \\"\\",
      \\"substituteLeaveType\\": null,
      \\"type\\": \\"Direct\\",
      \\"useEarlyLeaveReason\\": false,"
`;

exports[`reducer() MOBILE_APP/MODULES/ATTENDANCE/DAILY_REQUEST/UI/REQUESTS/DIRECT_REQUEST/UPDATE should not be executed if state is not initializing 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;
